gemrb/core/PathFinder.cpp

```diff
@@ -0,0 +1,382 @@
+{
+	std::vector<float> dist(Width * Height);
+	static const size_t DEGREES_OF_FREEDOM = 8;
+	static const char dx[DEGREES_OF_FREEDOM] = {1, 0, -1, 0, 1, 1, -1, -1};
+	static const char dy[DEGREES_OF_FREEDOM] = {0, 1, 0, -1, 1, -1, 1, -1};
+	SearchmapPoint smptFleeFrom = SearchmapPoint(d.x / 16, d.y / 12);
+	std::priority_queue<PQNode> open;
+	static const float diagWeight = sqrt(2);
+	dist[smptFleeFrom.y * Width + smptFleeFrom.x] = 0;
+	open.push(PQNode(smptFleeFrom, 0));
+	SearchmapPoint smptFarthest;
+	SearchmapPoint smptCurrent;
+	SearchmapPoint smptChild = smptFleeFrom;
+
+	float farthestDist = 0;
+	PQNode newNode = PQNode();
+	while (!open.empty()) {
+		smptCurrent = open.top().point;
+		open.pop();
+		// Randomly search clockwise or counterclockwise
+		// To prevent always fleeing in the same direction
+		int sgn = (RAND(0, 1)) ? -1 : 1;
+		for (size_t i = 0; i < DEGREES_OF_FREEDOM; i++) {
+			smptChild.x = smptCurrent.x + sgn * dx[i];
+			smptChild.y = smptCurrent.y + sgn * dy[i];
+			bool childOutsideMap =
+				smptChild.x <= 0 ||
+				smptChild.y <= 0 ||
+				(unsigned) smptChild.x >= Width ||
+				(unsigned) smptChild.y >= Height;
+			bool childBlocked = CheckSearchmapPointFlags(smptChild.x * 16 + 8, smptChild.y * 12 + 6, size,
+					validFlags, false);
+			if (!childBlocked && !childOutsideMap) {
+				float curDist = dist[smptCurrent.y * Width + smptCurrent.x];
+				float oldDist = dist[smptChild.y * Width + smptChild.x];
+				float newDist = curDist + (dx[i] && dy[i] ? diagWeight : 1);
+				if (newDist <= pathLength && newDist > oldDist) {
+					if (newDist > farthestDist) {
+						farthestDist = newDist;
+						smptFarthest = smptChild;
+					}
+					dist[smptChild.y * Width + smptChild.x] = newDist;
+					newNode.point.x = smptChild.x;
+					newNode.point.y = smptChild.y;
+					newNode.dist = newDist;
+					open.push(newNode);
+				}
+			}
+		}
+	}
+	return smptFarthest;
+}
+
+bool Map::TargetUnreachable(const Point &s, const Point &d, unsigned int size, bool actorsAreBlocking)
+{
+	int flags = PF_SIGHT;
+	if (actorsAreBlocking) { flags |= PF_ACTORS_ARE_BLOCKING; }
+	return FindPath(s, d, size, 0, flags) == NULL;
+}
+
+// Use this function when you target something by a straight line projectile (like a lightning bolt, arrow, etc)
+PathNode* Map::GetLine(const Point &start, const Point &dest, int flags)
+{
+	int Orientation = GetOrient(start, dest);
+	return GetLine(start, dest, 1, Orientation, flags);
+}
+
+PathNode* Map::GetLine(const Point &start, int Steps, int Orientation, int flags)
+{
+	Point dest=start;
+
+	double xoff, yoff, mult;
+	if (Orientation <= 4) {
+		xoff = -Orientation / 4.0;
+	} else if (Orientation <= 12) {
+		xoff = -1.0 + (Orientation - 4) / 4.0;
+	} else {
+		xoff = 1.0 - (Orientation - 12) / 4.0;
+	}
+
+	if (Orientation <= 8) {
+		yoff = 1.0 - Orientation / 4.0;
+	} else {
+		yoff = -1.0 + (Orientation - 8) / 4.0;
+	}
+
+	mult = 1.0 / (std::fabs(xoff) > std::fabs(yoff) ? std::fabs(xoff) : std::fabs(yoff));
```

Comment:
For later, since you just moved it, this can be simplified with std::max, saving one fabs call.